pkg/policy/l4.go

```diff
@@ -1019,9 +1050,44 @@ func addL4Filter(policyCtx PolicyContext,
 // key format: "port/proto"
 type L4PolicyMap map[string]*L4Filter
 
+type policyFeatures uint8
+
+const (
+	denyRules policyFeatures = 1 << iota
+	authRules
+
+	allFeatures policyFeatures = ^policyFeatures(0)
+)
+
+func (pf *policyFeatures) setFeature(feature policyFeatures) {
+	*pf |= feature
+}
+
+func (pf policyFeatures) contains(feature policyFeatures) bool {
+	return pf&feature != 0
+}
+
+type L4DirectionPolicy struct {
+	PortRules L4PolicyMap
+
+	// features tracks properties of PortRules to skip code when features are not used
+	features policyFeatures
```

Comment:
nit: I was under the impression that `L4DirectionPolicy` was an immutable definition of desired state. But the `policyFeatures` now seems like a mutable cache that gets updated when the filters in the policy get updated. I can't tell if this is a problem yet or not, but it feels like it could cause consistency issues at some point down the line. I don't have a counterproposal.

Reply:
Maybe one way to ease my concern here would be to process the need for deny / authentication during initial policy processing and creation of the L4DirectionPolicy instead of during `attach()`.

Reply:
`policyFeatures` are immutable everywhere after attach. We compute new filters for each new policy update, they are only ever updated during the policy update, when they need to be merged. So attach() marks the end of the initial policy processing. After this the `selectorPolicy` is immutable, gets shared for multiple endpoint policies that then make use of the (now immutable) `policyFeatures`. Or this is the intent, anyway!

Reply:
Ah, somehow I figured that `attach()` happens potentially multiple times and hence could cause the state to become unexpectedly mutable. But actually that may be a misunderstanding on my side.

Reply:
```
$ git grep "Attach("
pkg/policy/l4.go:func (l4 *L4Policy) Attach(ctx PolicyContext) {
pkg/policy/l4_test.go:  policy.Attach(testPolicyContext)
pkg/policy/repository.go:       calculatedPolicy.Attach(&policyCtx)
pkg/policy/resolve.go:func (p *selectorPolicy) Attach(ctx PolicyContext) {
pkg/policy/resolve.go:          p.L4Policy.Attach(ctx)
```

Seems like once per identity and each identity has a unique policy. Then as you state, selectorPolicy gets recalculated and atomically set with the expectation of immutability. So we should be good on this front?

Reply:
Yes :-)